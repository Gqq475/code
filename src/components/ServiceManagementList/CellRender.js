import React from 'react'
import classes from './ServiceManagementList.scss'

type Props = {
  value: String,
  data: Object
};

export class CellRender extends React.Component {
  props: Props;

  render () {
    console.log('value,', this.props.value, this.props.data)
    switch (this.props.value) {
      case 'none':
      case 'destroyed':
        // status idle
        return (<div className={classes.cn}>
          <div style={{'background': '#9B9B9B'}} className={classes.cell}>{'Idle'}</div>
        </div>)
      case 'initing':
      case 'inited':
      case 'starting':
        // status creating or confirm the create/edit button to trigger
        // should not enable create/edit button in specific service
        return (<div className={classes.cn}>
          <div style={{'background': '#7ED321'}} className={classes.cell}>{'Creating'}</div>
        </div>)
      case 'feeding':
      case 'feeded':
        // status in progress or confirm the load button to trigger
        // should not enable feed button in specific service
        return (<div className={classes.cn}>
          <div style={{'background': '#F5A623'}} className={classes.cell}>{'In progress'}</div>
          {this.props.data['progress']
            ? (<div className={`${classes.cell} ${classes.progress}`}>{this.props.data.progress}</div>)
            : (<div className={`${classes.cell} ${classes.progress}`}>0%</div>)
          }
        </div>)
      case 'destroying':
        // status in destroying or confirm the destroy button to trigger
        // should not enable destroy button in specific service
        return (<div className={classes.cn}>
          <div style={{'background': '#D0021B'}} className={classes.cell}>{'Destroying'}</div>
        </div>)
      case 'started':
      default:
        // status ready
        return (<div className={classes.cn}>
          <div style={{'background': '#4A90E2'}} className={classes.cell}>{'Ready'}</div>
        </div>)
    }
  }
}

export default CellRender
