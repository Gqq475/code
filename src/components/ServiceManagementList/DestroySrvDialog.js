import React from 'react'
import classes from './Dialog.scss'
import { Button, Dialog } from '@blueprintjs/core'
import { connect } from 'react-redux'
const qca = __QCA__

type Props = {
  kernelId: Number,
  isOpen: boolean,
  toggleDialog: Function,
  token: String
};

export class DestroySrvDialog extends React.Component {
  props: Props;

  constructor (props) {
    super(props)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleSubmit () {
    console.log('handle Destroy')
    let url = `${qca}/ojs/kernel/action`
    fetch(url, {
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
        'Authorization': this.props.token
      },
      method: 'POST',
      body: JSON.stringify({
        'kernelId': this.props.kernelId.toString(),
        'cmd': 'destroy'
      })
    })
      .then(res => res.json())
      .then(data => {

      })

    this.props.toggleDialog()
  }

  render () {
    return (
      <div className={classes['Dialog-container']}>
        <Dialog
          isOpen={this.props.isOpen}
          onClose={this.props.toggleDialog}
          className={classes.dialog}
        >
          <div className={classes.header}>
            <h6 className={classes.headerTitle}>提示</h6>
          </div>
          <div className={classes.dialogBody}>
            <br />
            <p className={classes.cn}>Destroy Service</p>
            <div className={classes.btnGroup}>
              <Button className={classes.btnCancel} onClick={this.props.toggleDialog}>
                {_.CANCEL}
              </Button>
              <Button className={`pt-button pt-intent-primary ${classes.btnEdit}`} onClick={this.handleSubmit}>
                {_.OK}
              </Button>
            </div>
          </div>
        </Dialog>
      </div>
    )
  }
}
export default connect(
  (state) => ({
    token: state.auth.token
  }),
  (dispatch, state) => {
    return {
      dispatch
    }
  }
)(DestroySrvDialog)
