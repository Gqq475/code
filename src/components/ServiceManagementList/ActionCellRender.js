import React from 'react'
import FeedImgDialog from './FeedImgDialog'
import DestroySrvDialog from './DestroySrvDialog'
import DeleteDialog from './DeleteDialog'
import classes from './ServiceManagementList.scss'
import OurToaster from 'components/OurToaster'
import { connect } from 'react-redux'
const qca = __QCA__

type Props = {
  value:String,
  data: Object,
  api: Object,
  token: String
};

export class ActionCellRender extends React.Component {
  props: Props;

  constructor (props) {
    super(props)
    this.onSetDataSrc = this.onSetDataSrc.bind(this)
    this.onDestroy = this.onDestroy.bind(this)
    this.onPause = this.onPause.bind(this)
    this.onDelete = this.onDelete.bind(this)
    this.state = {
      isFeedBtnOpen: false,
      isDestroyBtnOpen: false,
      isDeleteBtnOpen: false
    }
  }

  onPause () {
    console.log('执行暂停')
    fetch(`${qca}/ojs/kernel/stopFeed`,
      {
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
          'Authorization': this.props.token
        },
        method: 'POST',
        body: JSON.stringify({'kernelId': this.props.data.kernelId.toString()})
      }
    )
    .then(response => response.json())
    .then((data) => {
      if (data.code === 200000) {
        OurToaster.show({message: '暂停成功!'})
      } else {
        OurToaster.show({message: '暂停失败!'})
      }
    })
  }

  onSetDataSrc () {
    this.setState({
      isFeedBtnOpen: !this.state.isFeedBtnOpen
    })
  }

  onDestroy () {
    this.setState({
      isDestroyBtnOpen: !this.state.isDestroyBtnOpen
    })
  }

  onDelete () {
    this.setState({
      isDeleteBtnOpen: !this.state.isDeleteBtnOpen
    })
  }

  render () {
    switch (this.props.value) {
      case 'started':
        // status ready
        return <div className={classes.cn}>
          <button disabled className={'pt-button pt-minimal pt-icon-trash'} title='Delete' />
          <button disabled={false} onClick={this.onSetDataSrc}
            className={'pt-button pt-minimal pt-icon-document-open'}
            title='Feed image' />
          <button disabled={false} onClick={this.onDestroy}
            className={'pt-button pt-minimal pt-icon-delete'}
            title='Destroy' />
          <button disabled className={'pt-button pt-minimal pt-icon-history'} title='Log' />
          <FeedImgDialog id={this.props.data.id} deviceId={this.props.data.deviceId}
            kernelId={this.props.data.kernelId}
            modelId={this.props.data.modelId}
            isOpen={this.state.isFeedBtnOpen} toggleDialog={this.onSetDataSrc} />
          <DestroySrvDialog deviceId={this.props.data.deviceId}
            kernelId={this.props.data.kernelId}
            isOpen={this.state.isDestroyBtnOpen} toggleDialog={this.onDestroy} />
        </div>

      case 'feeding':
        return <div className={classes.cn}>
          <button disabled className={'pt-button pt-minimal pt-icon-trash'} title='Delete' />
          <button disabled={false}
            onClick={this.onPause}
            className={'pt-button pt-minimal pt-icon-pause'}
            title='Pause' />
          <button disabled={false} onClick={this.onDestroy}
            className={'pt-button pt-minimal pt-icon-delete'}
            title='Destroy' />
          <button disabled className={'pt-button pt-minimal pt-icon-history'} title='Log' />
          <FeedImgDialog id={this.props.data.id} deviceId={this.props.data.deviceId}
            kernelId={this.props.data.kernelId}
            modelId={this.props.data.modelId}
            isOpen={this.state.isFeedBtnOpen} toggleDialog={this.onSetDataSrc} />
          <DestroySrvDialog deviceId={this.props.data.deviceId}
            kernelId={this.props.data.kernelId}
            isOpen={this.state.isDestroyBtnOpen} toggleDialog={this.onDestroy} />
        </div>

      case 'none':
      case 'destroyed':
        return <div className={classes.cn}>
          <button disabled={false}
            className={'pt-button pt-minimal pt-icon-trash'}
            title='Delete'
            onClick={this.onDelete} />
          <button disabled onClick={this.onSetDataSrc}
            className={'pt-button pt-minimal pt-icon-document-open'}
            title='Feed image' />
          <button disabled onClick={this.onDestroy}
            className={'pt-button pt-minimal pt-icon-delete'}
            title='Destroy' />
          <button disabled className={'pt-button pt-minimal pt-icon-history'} title='Log' />
          <FeedImgDialog id={this.props.data.id} deviceId={this.props.data.deviceId}
            kernelId={this.props.data.kernelId}
            modelId={this.props.data.modelId}
            isOpen={this.state.isFeedBtnOpen} toggleDialog={this.onSetDataSrc} />
          <DestroySrvDialog deviceId={this.props.data.deviceId}
            kernelId={this.props.data.kernelId}
            isOpen={this.state.isDestroyBtnOpen} toggleDialog={this.onDestroy} />
          <DeleteDialog id={this.props.data.id} kernelId={this.props.data.kernelId}
            isOpen={this.state.isDeleteBtnOpen} toggleDialog={this.onDelete}
            api={this.props.api} />
        </div>

      default:
        return <div className={classes.cn}>
          <button disabled className={'pt-button pt-minimal pt-icon-trash'} title='Delete' />
          <button disabled onClick={this.onSetDataSrc}
            className={'pt-button pt-minimal pt-icon-document-open'}
            title='Feed image' />
          <button disabled onClick={this.onDestroy}
            className={'pt-button pt-minimal pt-icon-delete'}
            title='Destroy' />
          <button disabled className={'pt-button pt-minimal pt-icon-history'} title='Log' />
          <FeedImgDialog id={this.props.data.id} deviceId={this.props.data.deviceId}
            kernelId={this.props.data.kernelId}
            modelId={this.props.data.modelId}
            isOpen={this.state.isFeedBtnOpen} toggleDialog={this.onSetDataSrc} />
          <DestroySrvDialog deviceId={this.props.data.deviceId}
            kernelId={this.props.data.kernelId}
            isOpen={this.state.isDestroyBtnOpen} toggleDialog={this.onDestroy} />
        </div>
    }
  }
}
export default connect(
  (state) => ({
    token: state.auth.token
  }),
  (dispatch, state) => {
    return {
      dispatch
    }
  }
)(ActionCellRender)
