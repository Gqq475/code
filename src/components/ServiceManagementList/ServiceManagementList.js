import React from 'react'
import classes from './ServiceManagementList.scss'
import {AgGridReact} from 'ag-grid-react'
import CreateServiceDialog from 'components/CreateServiceDialog'
import CellRender from './CellRender'
import ActionCellRender from './ActionCellRender'
import Mqtt from 'mqtt'
import { connect } from 'react-redux'
import { identifyRole } from '../../Utils.js'
const qca = __QCA__

type Props = {
  data: Object,
  setDetail: Function,
  getModel: Function,
  getCreateService: Function,
  userRole: Array
};

export class ServiceManagementList extends React.Component {
  props: Props;
  constructor (props) {
    super(props)
    this.onRowClicked = this.onRowClicked.bind(this)
    this.datasource = this.datasource.bind(this)
    this.toggleDialog = this.toggleDialog.bind(this)
    this.onGridReady = this.onGridReady.bind(this)
    this.isVisibleStatusChanged = this.isVisibleStatusChanged.bind(this)
    this.onMsg = this.onMsg.bind(this)
    this.kernelIdArray = []

    this.msgQueue = []

    const columnDefs = [
      {headerName: 'Service ID', width: 90, field: 'id', suppressMovable: true, cellClass: classes.cn},
      {headerName: 'Service Name', field: 'name', suppressSorting: true, suppressMovable: true, cellClass: classes.cn},
      {headerName: 'Status-All',
        suppressMovable: true,
        field: 'status',
        cellRendererFramework: CellRender
      },
      {headerName: 'Model ID', width: 77, field: 'modelId', suppressMovable: true, cellClass: classes.cn},
      {headerName: 'Data Source', width: 150, field: 'dataSource', suppressMovable: true, cellClass: classes.cn},
      {headerName: 'Create Time',
        width: 180,
        field: 'createTime',
        sort: 'desc',
        suppressMovable: true,
        cellClass: classes.cn},
      {headerName: 'Action',
        suppressMovable: true,
        width: 180,
        field: 'status',
        suppressSorting: true,
        cellRendererFramework: ActionCellRender
      }
    ]

    this.state = {
      columnDefs,
      isOpen: false
    }
  }

  // WORKAROUND
  shouldComponentUpdate () {
    if (this.bypassRender) {
      this.bypassRender = false
      return false
    }

    return true
  }

  onRowClicked (e) {
    let target = e.event.target
    // ByPass Button click
    if (target !== undefined && target.nodeName !== 'BUTTON') {
      let data = e.data
      this.bypassRender = true
      this.props.setDetail(data)
    }
  }

  datasource () {
    let smIns = this
    return {
      rowCount: 1,
      getRows: (params) => {
        smIns.getRowsType = true

        let pageSize = 10
        let pageNow = params.endRow / pageSize
        const url = `${qca}/ojs/services?pageNum=${pageNow}&perPage=${pageSize}`
        fetch(url, {
          headers: {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
            'Authorization': smIns.props.token
          }
        })
        .then(response => response.json())
        .then((data) => {
          let kernelIdArray = []
          smIns.result = data.result.Services
          smIns.Count = data.result.Count
          let lastRow = smIns.Count
          let result = smIns.result

          result.map((unitResult) => {
            if (!kernelIdArray.includes(unitResult.kernelId)) {
              kernelIdArray.push(unitResult.kernelId)
            }
          })

          if (!smIns.client) {
            if (__PROD__) {
              smIns.client = Mqtt.connect(`wss://${location.hostname}:59001`, {
                'rejectUnauthorized': false,
                'username': 'deltaww',
                'password': 'Dare2Define'
              })
            } else {
              smIns.client = Mqtt.connect(__MQTT__, {
                'rejectUnauthorized': false,
                'username': 'deltaww',
                'password': 'Dare2Define'
              })
            }
            smIns.kernelIdArray = kernelIdArray
            let client = smIns.client
            client.on('message', function (topic, message) {
              smIns.onMsg && smIns.onMsg(topic, message)
              console.log('message', message)
            })
            kernelIdArray.map((kernelId) => {
              client.on('connect', function () {
                client.subscribe(`autojudge/${kernelId}/agent`)
              })
            })
          } else {
            kernelIdArray.forEach((kernelId) => {
              if (!smIns.kernelIdArray.includes(kernelId)) {
                smIns.client.subscribe(`autojudge/${kernelId}/agent`)
              }
            })
            smIns.kernelIdArray = kernelIdArray.concat()
          }

          params.successCallback(result, lastRow)
          /* while (smIns.msgQueue.length) {
            smIns.getRowsType && (smIns.getRowsType = false)
            const item = smIns.msgQueue.pop()
            smIns.onMsg(item.topic, item.message)
          } */
          smIns.getRowsType = false
        })
      }
    }
  }

  onMsg (topic, message) {
    console.log('topic', topic, JSON.parse(message.toString()))
    if (!(/autojudge\/\d{1,}\/agent/).exec(topic)) {
      console.log('topic', topic, message)
      return
    }

    // this.getRowsType===trueï¼Œ Delayed update
    /* if (this.getRowsType) {
      this.msgQueue.push({topic, message})
      return
    } */

    let id = topic.match(/\/(.*?)\//)[1]
    let msgArray = JSON.parse(message.toString())
    if (msgArray['cmd'] !== 'status' && msgArray['cmd'] !== 'progress') {
      return
    }

    let updatedNodes = []
    let isChanged = false

    // Status updates
    msgArray['cmd'] === 'status' && this.api.forEachNode((node) => {
      let unitResult = node.data
      if (unitResult.kernelId.toString() === id &&
        msgArray['reason'] === '' &&
        this.isVisibleStatusChanged(unitResult.status, msgArray['payload'])) {
        unitResult.status = msgArray['payload']
        isChanged = true
      }
      node.data.progress = 0
      updatedNodes.push(node)
    })

    // Add a progress bar
    msgArray['cmd'] === 'progress' && this.api.forEachNode((node) => {
      console.log('progress')
      node.data.progress = ((+msgArray['payload']).toFixed(3) * 100) + '%'
      isChanged = true
      updatedNodes.push(node)
    })

    isChanged && this.api.refreshRows(updatedNodes)
  }

  isVisibleStatusChanged (oldStatus, newStatus) {
    return (this.getVisibleStatusCode(oldStatus) !== this.getVisibleStatusCode(newStatus)) ||
    ((oldStatus === 'started') !== (newStatus === 'started'))
  }

  getVisibleStatusCode (status) {
    switch (status) {
      case 'none':
      case 'destroyed':
        return 0
      case 'initing':
      case 'inited':
      case 'starting':
        return 1
      case 'started':
        return 2
      case 'feeding':
      case 'feeded':
        return 3
      case 'destroying':
        return 4
      default:
        return -1
    }
  }

  onGridReady (params) {
    this.api = params.api
    // this.api.setDatasource(this.datasource())
  }

  toggleDialog () {
    this.setState({
      isOpen: !this.state.isOpen
    })
  }

  componentWillUnmount () {
    // UnSubscribe mqtt
    this.client && this.client.end(() => {
      console.log('Disconnect')
    })
  }

  render () {
    return (
      <div className={classes['ServiceManagementList-container']}>
        <section className={`${classes.title} ${classes.clear}`}>
          <h4 className={`${classes.left}`}>Service List
            {
              (identifyRole(this.props.userRole) === 'admin') ? (
                <button type='button' onClick={this.toggleDialog}
                  className={`${classes.ml10} pt-button pt-intent-primary pt-icon-plus`} />
              ) : null
            }
            {
              this.state.isOpen
              ? <CreateServiceDialog isOpen={this.state.isOpen} toggleDialog={this.toggleDialog}
                data={this.props.data}
                getModel={this.props.getModel}
                getCreateService={this.props.getCreateService}
                dataList={this.result} />
              : ''
            }
          </h4>
        </section>

        <section className={`${classes['ag-table']} ag-fresh`}>
          <AgGridReact
            enableSorting
            sortingOrder={['desc', 'asc']}
            rowSelection='single'
            rowModelType='pagination'
            paginationPageSize={10}
            onGridReady={this.onGridReady}
            columnDefs={this.state.columnDefs}
            onRowClicked={this.onRowClicked}
            datasource={this.datasource()}
          />
        </section>
      </div>
    )
  }
}
export default connect(
  (state) => ({
    token: state.auth.token,
    userRole: state.auth.userRole
  }),
  (dispatch, state) => {
    return {
      dispatch
    }
  }
)(ServiceManagementList)
