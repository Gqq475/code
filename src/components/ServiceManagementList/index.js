import ServiceManagementList from './ServiceManagementList'
export default ServiceManagementList
