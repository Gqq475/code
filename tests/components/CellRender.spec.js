import React from 'react'
import { CellRender } from 'components/ServiceManagementList/CellRender'

describe('(Component) CellRender', () => {
  it('render normally', () => {
    const wrapper = shallow(
      <CellRender />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
