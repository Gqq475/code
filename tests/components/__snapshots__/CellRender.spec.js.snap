// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) CellRender render normally 1`] = `
<div
  className="cn"
>
  <div
    className="cell"
    style={
      Object {
        "background": "#4A90E2",
      }
    }
  >
    Ready
  </div>
</div>
`;
